js/private/js_run_devserver.js

```diff
@@ -0,0 +1,209 @@
+const path = require('path')
+const perf_hooks = require('perf_hooks')
+const fs = require('fs')
+const os = require('os')
+const child_process = require('child_process')
+
+// Globals
+const RUNFILES_ROOT = path.join(
+    process.env.JS_BINARY__RUNFILES,
+    process.env.JS_BINARY__WORKSPACE
+)
+const synced = new Map()
+const mkdirs = new Set()
+
+// Returns true if a file's stats indicate it is an executable file.
+function isExecutable(stats) {
+    return (
+        !!(stats.mode & fs.constants.S_IXUSR) ||
+        !!(stats.mode & fs.constants.S_IXGRP) ||
+        !!(stats.mode & fs.constants.S_IXOTH)
+    )
+}
+
+// Ensure that a directory exists. If it has not been previously created or does not exist then it
+// creates the directory, first recursively ensuring that its parent directory exists.
+// Intentionally synchronous to avoid race conditions between sync promises
```

Comment:
*async ?

Any examples of such race conditions that could be mentioned?

Reply:
When you use `await fs.promises.mkdir(p)` then you could end up calling it twice in two different promises which would error on the 2nd call. This is because `if (!fs.existsSync(p))` followed by `await fs.promises.mkdir(p)` is not atomic so both threads can enter into the condition before either calls mkdir.

Reply:
Updated comment